### PR TITLE
aboutUrl function

### PR DESCRIPTION
**Pull request by CharlesRendle**
aboutUrl function which places multimeasure dimensions at the end of the URL with two accompanying unit tests.

Have added a todo regarding a future least -> most entropic dimension ordering approach.

### CODE REVIEW

File: csvqb/csvqb/writers/qbwriter.py [hunk 1]
```diff
@@ -421,3 +422,25 @@ def _get_measure_uri(measure: QbMeasure) -> str:
...
+def _get_about_url(cube: Cube) -> str:
+    # Todo: Dimensions are currently appended in the order in which the appear in the cube.
+    #       We may want to alter this in the future so that the ordering is from
+    #       least entropic dimension -> most entropic.
+    #       E.g. http://base-uri/observations/male/1996/all-males-1996
+    multi_measure_cols = []
+    aboutUrl = "#obs"
+    for c in cube.columns:
+        if isinstance(c, QbColumn):
+            c: QbColumn
```
Comment (robons) [inline csvqb/csvqb/writers/qbwriter.py:436]: I think this is [breaking the build](https://ci.floop.org.uk/job/GSS_data/job/csvwlib/view/change-requests/job/PR-98/1/console).

> /var/jenkins_home/workspace/GSS_data_csvwlib_PR-98/csvqb/csvqb/writers/qbwriter.py:434:9 - error: Expression of type "CsvColumn" cannot be assigned to declared type "QbColumn[Unknown]"

File: csvqb/csvqb/tests/unit/cube/qb/errorvalidation.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def test_single_measure_qb_definition(self):
             "Value": [2, 2, 2]
         })
 
-        metadata = CubeMetadata(URI("http://example.com/some/dataset"), "Some Dataset")
+        metadata = CubeMetadata("Some Dataset")
```
Comment (robons) [inline csvqb/csvqb/tests/unit/cube/qb/errorvalidation.py:22]: Can you revert this change? I've got a fix for it in another PR and don't want more conflicts than we absolutely need.

File: csvqb/csvqb/tests/unit/writers/qbwritertests.py [hunk 3]
```diff
@@ -303,6 +303,66 @@ def test_virtual_columns_generated_for_multi_meas_obs_val(self):
...
+            QbColumn("Value", 
+                     QbSingleMeasureObservationValue(ExistingQbMeasure("http://example.com/measures/existing_measure"), 
+                     NewQbUnit("New Unit")))
+            
+        ]
+
+        cube = Cube(metadata, data, columns)
+
+        actual_about_url = qbwriter._get_about_url(cube)
+        expected_about_url = "#obs/{+existing-dimension}/{+local-dimension}"
```
Comment (robons) [inline csvqb/csvqb/tests/unit/writers/qbwritertests.py:333]: hmm, these should be underscores, not dashes. Something doesn't seem quite right here.

File: csvqb/csvqb/writers/qbwriter.py [hunk 4]
```diff
@@ -421,3 +422,25 @@ def _get_measure_uri(measure: QbMeasure) -> str:
...
+            if isinstance(c.component, QbDimension):
+                aboutUrl = aboutUrl + f"/{{+{c.uri_safe_identifier}}}"
```
Comment (robons) [inline csvqb/csvqb/writers/qbwriter.py:437]: Ah, I see why you have dashes, we need to use the [csvw_column_name_safe](https://github.com/GSS-Cogs/csvwlib/blob/7f8ce259d2dc04f3b8dec9aae1a9c20de0f0ce2a/csvqb/csvqb/writers/qbwriter.py#L258) function to be consistent with the names defined in the CSV-W columns section.

File: csvqb/csvqb/writers/qbwriter.py [hunk 5]
```diff
@@ -421,3 +422,25 @@ def _get_measure_uri(measure: QbMeasure) -> str:
...
+            elif isinstance(c.component, QbMultiMeasureDimension):
+                multi_measure_cols.append(c.uri_safe_identifier)
+            else: 
+                pass
+    if len(multi_measure_cols) != 0:
+        for m in multi_measure_cols:
```
Comment (robons) [inline csvqb/csvqb/writers/qbwriter.py:443]: There should only ever by *one* multi measure column. So this shouldn't be a list.

**Comment by CharlesRendle** [comment]
All changes made based on PR comments

File: csvqb/csvqb/tests/unit/writers/qbwritertests.py [hunk 6]
```diff
@@ -303,6 +303,66 @@ def test_virtual_columns_generated_for_multi_meas_obs_val(self):
...
+            "Measure": ["People", "Children", "Adults"],
+            "Units": ["Percent", "People", "People"]
+        })
+
+        metadata = CubeMetadata("Some Dataset")
+        columns = [
+            QbColumn("Existing Dimension", ExistingQbDimension("https://example.org/dimensions/existing_dimension")),
+            QbColumn("Local Dimension", NewQbDimension.from_data("Name of New Dimension", data["Local Dimension"])),
+            QbColumn("Value", QbMultiMeasureObservationValue("number")),
+            QbColumn("Measure", QbMultiMeasureDimension.new_measures_from_data(data["Measure"])),
```
Comment (robons) [inline csvqb/csvqb/tests/unit/writers/qbwritertests.py:355]: Can you move the measure so that it's defined before the dimensions. We want to test that the measure ends up on the end of the URL even when it isn't defined last.

File: csvqb/csvqb/writers/qbwriter.py [hunk 7]
```diff
@@ -421,3 +422,23 @@ def _get_measure_uri(measure: QbMeasure) -> str:
...
+    #       E.g. http://base-uri/observations/male/1996/all-males-1996
+    aboutUrl = "#obs"
+    multi_measure_col = ""
+    for c in cube.columns:
+        if isinstance(c, QbColumn):
+            if isinstance(c.component, QbDimension):
+                aboutUrl = aboutUrl + f"/{{+{csvw_column_name_safe(c.uri_safe_identifier)}}}"
+            elif isinstance(c.component, QbMultiMeasureDimension):
+                multi_measure_col = csvw_column_name_safe(c.uri_safe_identifier)
+            else:
```
Comment (robons) [inline csvqb/csvqb/writers/qbwriter.py:440]: You don't need this else statement here, may as well remove it.

**Review by robons — APPROVED** [review]
You just need to fix your file conflicts before merging.

File: csvqb/csvqb/tests/unit/writers/qbwritertests.py [hunk 8]
```diff
@@ -485,6 +485,66 @@ def test_virtual_columns_generated_for_multi_meas_obs_val(self):
...
+            QbColumn("Value", 
+                     QbSingleMeasureObservationValue(ExistingQbMeasure("http://example.com/measures/existing_measure"), 
+                     NewQbUnit("New Unit")))
+            
+        ]
+
+        cube = Cube(metadata, data, columns)
+
+        actual_about_url = qbwriter._get_about_url(cube)
+        expected_about_url = "#obs/{+existing_dimension}/{+local_dimension}"
```
Comment (robons) [inline csvqb/csvqb/tests/unit/writers/qbwritertests.py:515]: I forgot to commit the changes we made in the test file! The tests will surely fail.

**Review by canwaf — CHANGES_REQUESTED** [review]
Blocking review for now.

**Review by canwaf — APPROVED** [review]

**Comment by canwaf** [comment]
Happy to see this merged now.
